### PR TITLE
[`FA2`] Cast to correct dtype

### LINKED ISSUE #26451: The hidden states in LlamaFlashAttention2 are cast in fp16 unexpectedly
### System Info

- `transformers` version: 4.33.1
- Platform: Linux-5.4.0-147-generic-x86_64-with-glibc2.31
- Python version: 3.10.13
- Huggingface_hub version: 0.17.1
- Safetensors version: 0.3.3
- Accelerate version: 0.23.0
- PyTorch version (GPU?): 2.0.1+cu118 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: A100 40GB
- Using distributed or parallel set-up in script?: No

### Who can help?

@younesbelkada 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

As we discussed in this thread: https://github.com/huggingface/transformers/pull/25598#discussion_r1338877983

The hidden states may be cast in float16 even if we are using bf16 mixed precision training.

https://github.com/huggingface/transformers/blob/78dd1202823ca035b9609ddbcdaac2945a6530ff/src/transformers/models/llama/modeling_llama.py#L485-L487

It may be difficult to figure out the correct data type if the model is loaded in 4/8-bit mode.

### Expected behavior

The hidden states should be cast in Bfloat16 in bf16 training.

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

Fixes: https://github.com/huggingface/transformers/issues/26451

Currently performing bf16 fine-tuning with FA-2 leads to hidden states silently being casted in float16
As it is challenging to retrieve the original dtype of the model in case the model is quantized, I propose to store that dtype in a private attribute to be able to retrieve it conveniently without having to perform any sort of hack that gets the correct dtype if the model is quantized

### CODE REVIEW

File: src/transformers/models/falcon/modeling_falcon.py [hunk 1]
```diff
@@ -613,15 +613,20 @@ def forward(
         # cast them back in float16 just to be sure everything works as expected.
         input_dtype = query_layer.dtype
         if input_dtype == torch.float32:
+            if getattr(self.config, "_flash_attn_2_attention_dtype", None) is not None:
+                attention_dtype = self.config._flash_attn_2_attention_dtype
+            else:
+                attention_dtype = torch.float16
```
Comment (ArthurZucker) [inline src/transformers/models/falcon/modeling_falcon.py:619]: maybe default in the modeling utils no?

**Review by ArthurZucker — COMMENTED** [review]
A small nit!

**Review by ArthurZucker — COMMENTED** [review]
Okay! LGTM but the issue is that the dtype could/might change when we do `model.to(device)` meaning this only fixes inference after init if `torch_dtype` is specified

**Review by ArthurZucker — COMMENTED** [review]
Changes to the modeling utils are a bit too specific (specifically the `to`)  to flash attention. Given that `torch_dtype` can be accessed in the `XXXFlashAttention` as `self.config` makes more sense to have this in the attention (if possible?).

It's good if we want this, but there might be a solution to just change the attention

**Comment by younesbelkada** [comment]
I agree we should make all hacks / changes with respect to FA2 modules inside them. However this might introduce multiple patches and other hacks for quantized models. I think for now this approach is fine, but I agree we should go for a better one, as this would unblock some users for the next release, I left it as a TODO!

File: src/transformers/models/mistral/modeling_mistral.py [hunk 2]
```diff
@@ -411,15 +411,19 @@ def forward(
...
+            else:
+                attention_dtype = torch.float16
+
             logger.warning_once(
-                "The input hidden states seems to be silently casted in float32, this might be related to"
-                " the fact you have upcasted embedding or layer norm layers in float32. We will cast back the input in"
-                " float16."
+                f"The input hidden states seems to be silently casted in float32, this might be related to"
+                f" the fact you have upcasted embedding or layer norm layers in float32. We will cast back the input in"
+                f" {attention_dtype}. Make sure to pass the desired dtype when calling `from_pretrained` with `torch_dtype=your_dtype`"
```
Comment (MrTimmy89) [inline src/transformers/models/mistral/modeling_mistral.py:419]: Seems to me, that this line is 124 long and should be split into 2, previous lines could be left without 'f' at the beginning, while there are no values to be formatted there

**Comment by younesbelkada** [comment]
cc @hiyouga are you able to Fine-tune in bf16 with this branch?

File: src/transformers/modeling_utils.py [hunk 3]
```diff
@@ -2172,6 +2175,26 @@ def to(self, *args, **kwargs):
...
+                        if isinstance(arg, torch.dtype):
+                            target_dtype = arg
+                            break
+                else:
+                    target_dtype = kwargs["dtype"]
+
+                if target_dtype is not None and target_dtype == torch.float32:
+                    raise ValueError(
+                        "You cannot cast a model that has been loaded with Flash Attention 2 in `float32`"
+                    )
```
Comment (LysandreJik) [inline src/transformers/modeling_utils.py:2194]: Maybe mention how to go around this?

File: src/transformers/modeling_utils.py [hunk 4]
```diff
@@ -2172,6 +2175,26 @@ def to(self, *args, **kwargs):
                 " model has already been set to the correct devices and casted to the correct `dtype`."
             )
         else:
+            # TODO: @younesbelkada find a better way to do this directly in `xxxFlashAttention` modules
+            # currently it is not possible to retrieve the original dtype for quantized models.
```
Comment (LysandreJik) [inline src/transformers/modeling_utils.py:2179]: Should this be investigated now?

**Review by LysandreJik — COMMENTED** [review]
Ok, cool to relax the cast to fp16. 

Wouldn't it be cleaner to put the changes relative to `to` in a `FlashAttention`-specific mixin so as to not require changes to `modeling_utils.py`? You wouldn't need to add an additional private property `_flash_attn_2_attention_dtype` and wouldn't need to edit the general `to` method either (just the Flash-attention-specific `to` method. 

You may want to update the `_apply` method instead of `to` however, I think with `to` you're not seeing calls like `model.float()` which will convert your entire model to float32.

And probably an edge case but with a class-level `to` override like this you're not keeping the following in check:
```
model = MistralForCausalLM.from_pretrained("hf-internal-testing/tiny-random-MistralModel", use_flash_attention_2=True, torch_dtype=torch.float16)
model.model.layers.to(torch.float32)
```

**Comment by hiyouga** [comment]
I also thought that we could retrieve the data type from self.config.torch_dtype

File: src/transformers/models/llama/modeling_llama.py [hunk 5]
```diff
@@ -476,15 +476,17 @@ def forward(
         # in fp32. (LlamaRMSNorm handles it correctly)
         input_dtype = query_states.dtype
         if input_dtype == torch.float32:
+            attention_dtype = self.config._flash_attn_2_attention_dtype
```
Comment (patrickvonplaten) [inline src/transformers/models/llama/modeling_llama.py:479]: Not all attention modules have a config variable. But guess that's ok we can just pass it forward.

**Comment by younesbelkada** [comment]
Closing this PR in favor of https://github.com/huggingface/transformers/pull/26846

**Pull request closed without merging**